### PR TITLE
Explain that hidden contents should not be referenced.

### LINKED ISSUE #3049: Has the Access.hide annotation a meaning for instances of the class?
If I have a class with annotation(Protection(access = Access.hide)); and an instance of this class in a model.
What I should be able to do and to see with this instance when I look at the model in diagram view?

According to the specification all kind of access to the class not possible, but applies this also to the instance?
- Is the instance visible at all?
- If yes, can I have access to public parameters? (this is what D*a does)
- ...

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #3049
Note that it is slightly different from the issue-text.
The key part is simply: don't use a class in such a way that it the class-reference can be seen if the corresponding class-contents cannot be shown as that doesn't make sense.

### CODE REVIEW

**Review by maltelenz — COMMENTED** [review]
I have to little real world experience using the higher level of protections to give good feedback on this.

File: chapters/annotations.tex [hunk 1]
```diff
@@ -2052,24 +2055,24 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
-The \lstinline!access! annotation holds for the respective class and all classes
-that are hierarchically on a lower level, unless overridden by a
-\lstinline!Protection! annotation with \lstinline!access!.
-Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate!
-has no effect.
+The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
+Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.
+
+Classes should not use other classes in ways that contradict this protection.
+For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2062]: There should be no room for interpretation:
```suggestion
Classes must not use other classes in ways that contradict this protection.
For instance, a class with \lstinline!Access.hide! must not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there must not be hyperlinks to classes with \lstinline!Access.icon! from classes with visible documentation.
```

Comment (henrikt-ma) [reply chapters/annotations.tex:2062]: I wonder if we could go even further, and give the explicit rules that the generic formulation boils down to?
```suggestion
Classes must not use other classes in ways that contradict this protection:
\begin{itemize}
\item A class with \lstinline!Access.hide! or higher protection must not be used in the diagram layer of a class with \lstinline!Access.diagram!.
\item There must not be hyperlinks to classes with \lstinline!Access.icon! or higher protection from classes with visible documentation.
…
\end{itemize}
```

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
One way or another, I think there should be less room for interpretation.  My preference would be to have a concrete set of rules rather than something generic with examples.

File: chapters/annotations.tex (hunk 1 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2062]: I feared that the full list would be rather long, which would risk that something was missed and thus it was clearer to just give the goal.

The use of "should" instead of "must" was deliberate. If a package makes an error here that should not prevent us from simulating its models, so a strict requirement could be a problem.

Comment (henrikt-ma) [reply chapters/annotations.tex:2062]: As a library author defending for my intellectual property, I'd be disappointed if the specification didn't strictly require tools to obey access restrictions.  If they aren't set up correctly, the library should be fixed so that there is no ambiguity about how a tool should handle the contents.

Comment (HansOlsson) [reply chapters/annotations.tex:2062]: I agree that tools should obey the access restrictions, but the above is stated as restrictions on the classes - not on the tools.

So, to clarify add 
"Tools must ensure that protected contents is not shown, even if classes do not meet this requirement."

To me this "must" only makes sense if the restriction for the classes use "should", since "must" is an absolute requirement and here we are specifying what happens if the requirement is broken.

Note that this doesn't describe how the situation is handled by tools as long as the contents isn't shown - for diagrams Dymola have depending on circumstances used either Crossed-out-icons and just no indication; but the classes with those diagrams could still be used. (Obviously the specific libraries were corrected later.)

Comment (henrikt-ma) [reply chapters/annotations.tex:2062]: Sounds like a good discussion point for a web meeting.

Comment (HansOlsson) [reply chapters/annotations.tex]: This was now added after the web-meeting.

File: chapters/annotations.tex [hunk 2]
```diff
@@ -2052,24 +2055,26 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
-The \lstinline!access! annotation holds for the respective class and all classes
-that are hierarchically on a lower level, unless overridden by a
-\lstinline!Protection! annotation with \lstinline!access!.
-Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate!
-has no effect.
+The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
+Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.
+
+Classes should not use other classes in ways that contradict this protection.
+Tools must ensure that protected contents is not shown, even if classes do not meet this requirement.
```
Comment (beutlich) [inline chapters/annotations.tex:2062]:
```suggestion
Tools must ensure that protected content is not shown, even if classes do not meet this requirement.
```

File: chapters/annotations.tex [hunk 3]
```diff
@@ -2052,24 +2055,26 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+
+For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).
```
Comment (beutlich) [inline chapters/annotations.tex:2064]: Have the examples non-normative?

**Review by beutlich — CHANGES_REQUESTED** [review]

File: chapters/annotations.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2062]: I'm not sure if that change is necessary, but I'm not totally against it.

"Contents" implies that it is more tangible and countable, and to me the individual elements are so tangible and countable and none of them should be shown.

However, I can understand "content" in the sense that it is the protected ideas represented there that shouldn't be shown.

File: chapters/annotations.tex (hunk 3 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2064]: Could possibly do, but there has also been a push for making it more normative.

File: chapters/annotations.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/annotations.tex]: Yes. Done.

File: chapters/annotations.tex (hunk 3 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2065]: Added now.

File: chapters/annotations.tex [hunk 4]
```diff
@@ -2052,24 +2055,28 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
-The \lstinline!access! annotation holds for the respective class and all classes
-that are hierarchically on a lower level, unless overridden by a
-\lstinline!Protection! annotation with \lstinline!access!.
-Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate!
-has no effect.
+The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
+Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.
+
+Classes should not use other classes in ways that contradict this protection.
+Tools must ensure that protected contents are not shown, even if classes do not meet this requirement.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2062]: Stronger:
```suggestion
A class may only be used in ways that comply with this protection.
Tools must ensure that protected contents are not revealed, even in case this rule about valid use is violated.
```

File: chapters/annotations.tex [hunk 5]
```diff
@@ -2052,24 +2055,28 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+
+\begin{nonnormative}
+For instance a class with \lstinline!Access.hide! should not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there should not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2065]: Stronger:
```suggestion
For instance, a class with \lstinline!Access.hide! must not be used in the diagram layer of a class with \lstinline!Access.diagram!, and there must not be hyperlinks to classes with \lstinline!Access.icon! (from classes with visible documentation).
```

File: chapters/annotations.tex [hunk 6]
```diff
@@ -2052,24 +2055,28 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+\end{nonnormative}
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2066]: The rule about what tools should do when the usage rule has been violated would also deserve some illustration.
```suggestion
\end{nonnormative}

\begin{example}
Consider the following invalid use of a class with \lstinline!Access.hide!:
\begin{lstlisting}[language=modelica]
block MySecret
  RealOutput y = time;
  annotation(Protection(access = Access.hide));
end MySecret;

model M
  MySecret mySecret;
  Integrator integrator;
equation
  connect(mySecret.y, integrator.u);
  annotation(Protection(access = Access.diagram));
end M;
\end{lstlisting}
In order to not reveal the existence of the class \lstinline!MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.
(The tool could also choose to not show the diagram of \lstinline!M! at all.)
When simulating \lstinline!M!, the tool must not store \lstinline!mySecret.y!

Had \lstinline!M! had \lstinline!Access.packageDuplicate! instead, one way for a tool to not reveal the existence of \lstinline!MySecret! would be to refuse to show the class definition of \lstinline!M! at all.
\end{example}
```

**Review by henrikt-ma — DISMISSED** [review]
I don't like the rule that tools are held responsible for not revealing information when protection usage restrictions have been violated.  I know others have a different opinion here, so I'm just suggesting that an example is added to show the consequences of this rule more clearly.

**Comment by d-hedberg** [comment]
If a tool allows libraries to be produced where protection usage restrictions have been violated, these libraries must be considered corrupt.

The only sensible and robust way of dealing with this by other tools (that do have a mechanism in place to detect and prevent this violation) would be to not allow such a library to load at all. I fail to see why the responsibility to deal with such a corrupt library produced by one tool should fall on the other tools (more than to inform the user that the library violates the protection usage restrictions and cannot be permitted to load).

**Comment by henrikt-ma** [comment]
> > The only sensible and robust way of dealing with this by other tools (that do have a mechanism in place to detect and prevent this violation) would be to not allow such a library to load at all. I fail to see why the responsibility to deal with such a corrupt library produced by one tool should fall on the other tools (more than to inform the user that the library violates the protection usage restrictions and cannot be permitted to load).
> 
> I find that a bit harsh - but it is consistent with not exposing the wrong contents.

I like the simplicity of this approach, and I think it would serve well as an example in the specification of how a tool may interpret the requirement – better than the examples I've suggested [above](https://github.com/modelica/ModelicaSpecification/pull/3232#discussion_r986237271).

**Comment by henrikt-ma** [comment]
Another weird situation for a tool to handle is when `EncryptedLib.A` has `Access.hide`, and the user enters the text of this top-level model:
{code}
model M
  EncryptedLib.A a;
end M;
{code}

Should the tool suddenly close the text view and just show the icon of `M`?  As I see it, it is not reasonable to require that a tool takes action to not reveal the existence of `EncryptedLib.A` when this happens.  Instead, I think the tool's responsibility should be to give an error when flattening `M`, and also refuse to save `M` to permanent storage.

File: chapters/annotations.tex (hunk 6 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2066]: I believe it is important to show that M and MySecret are part of the same encrypted package, so enclose them in package P (no need to have a longer name). For the final sentence I would just completely remove it or reformulate it based on the other example:

If a user entered
```
model My
  P.MySecret a;
end My;
```
That should fail similarly as if `P.MySecret` did not exist, and especially the icon for `P.MySecret` should not be used.

-
Basically a class may be revealed in Modelica text, in documentation, in comments, or just guessed. As soon as we say that a hidden class is different from a non-existent class we open up a way to reveal their existence.

Thus trying to force tools to treat hidden classes in badly designed package specially we would force tools to reveal the existence of hidden classes in normal cases; that doesn't make sense.

Comment (henrikt-ma) [reply chapters/annotations.tex:2066]: > That should fail similarly as if P.MySecret did not exist, and especially the icon for P.MySecret should not be used.

This makes sense.  I'll use it to reformulate the example.

File: chapters/annotations.tex [hunk 7]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+package P
+  block MySecret
+    RealOutput y = time;
+    annotation(Protection(access = Access.hide));
+  end MySecret;
+  model M
+    MySecret mySecret;
+    Integrator integrator;
+  equation
+    connect(mySecret.y, integrator.u);
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2082]: To fully illustrate the point of the example, one should also try to place the icons on the diagram:
```suggestion
    MySecret mySecret
     annotation(Placement(
       transformation(origin = {30, 30}, extent = {{-10, -10}, {10, 10}})));
    Integrator integrator
     annotation(Placement(
       transformation(origin = {70, 30}, extent = {{-10, -10}, {10, 10}})));
  equation
    connect(mySecret.y, integrator.u)
      annotation(Line(origin = {49.5, 30}, points = {{-8.5, 0}, {8.5, -0}}));
```

File: chapters/annotations.tex [hunk 8]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+    annotation(Protection(access = Access.diagram));
+  end M;
+end P;
+\end{lstlisting}
+In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.
+(The tool could also choose to not show the diagram of \lstinline!P.M! at all.)
+When simulating \lstinline!P.M!, the tool must not store \lstinline!mySecret.y!
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2089]: Punctuation:
```suggestion
When simulating \lstinline!P.M!, the tool must not store \lstinline!mySecret.y!.
```

File: chapters/annotations.tex [hunk 9]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+    MySecret mySecret;
+    Integrator integrator;
+  equation
+    connect(mySecret.y, integrator.u);
+    annotation(Protection(access = Access.diagram));
+  end M;
+end P;
+\end{lstlisting}
+In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.
+(The tool could also choose to not show the diagram of \lstinline!P.M! at all.)
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2088]: Possibly to be combined with the sentence on the next list:
```suggestion
As long as the invalid use of \lstinline!P.MySecret! occurs within the same top level package as where the class is defined (here, \lstinline!P!), a tool is allowed to silently ignore the use for purposes of model translation.
```

File: chapters/annotations.tex [hunk 10]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+When simulating \lstinline!P.M!, the tool must not store \lstinline!mySecret.y!
+
+If a user enters
+\begin{lstlisting}[language=modelica]
+model My
+  P.MySecret a;
+end My;
+\end{lstlisting}
+That icon for \lstinline!P.MySecret! shall not be shown, similarly as if the class did not exist.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2097]: Make this a separate example with focus on lookup behavior:
```suggestion
In order to support development of valid protected packages, it is of course OK and expected that a tool will report the invalid use of \lstinline!P.MySecret! in \lstinline!P.M! (revealing its existence in a diagnostic) during development of the package.
\end{example}

\begin{example}
With the same package \lstinline!P! as in the previous example, consider invalid outside of \lstinline!P!:
\begin{lstlisting}[language=modelica]
model My
  P.MySecret a
    annotation(Placement(
       transformation(origin = {30, 30}, extent = {{-10, -10}, {10, 10}})));;
end My;
\end{lstlisting}
Regardless of the protection of \lstinline!My!, a tool must act as if \lstinline!P.MySecret! did not exist.
For example, translation of \lstinline!My! must fail with a standard error message about reference to the non-existing class \lstinline!P.MySecret!.
Further, just like when being misused inside the package \lstinline!P!, the tool must not reveal that it knows about the icon of \lstinline!P.MySecret!.
With such precautions taken, showing the text or diagram of \lstinline!My! is permitted as it doesn't reveal the actual existence of \lstinline!P.MySecret!.
```

File: chapters/annotations.tex [hunk 11]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+    annotation(Protection(access = Access.hide));
+  end MySecret;
+  model M
+    MySecret mySecret;
+    Integrator integrator;
+  equation
+    connect(mySecret.y, integrator.u);
+    annotation(Protection(access = Access.diagram));
+  end M;
+end P;
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2085]: The problematic case of even less protection within the package should also be discussed:
```suggestion
  model M2
    MySecret mySecret;
    annotation(Protection(access = Access.packageDuplicate));
  end M2;
end P;
```

File: chapters/annotations.tex [hunk 12]
```diff
@@ -2052,24 +2055,56 @@ \subsection{Protection of Classes}\label{protection-of-classes}
...
+\end{lstlisting}
+In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram with both \lstinline!mySecret! and all connections to it removed.
```
Comment (henrikt-ma) [inline chapters/annotations.tex:2087]: Clarify which diagram:
```suggestion
In order to not reveal the existence of the class \lstinline!P.MySecret!, a tool may choose to show the diagram of \lstinline!P.M! with both \lstinline!mySecret! and all connections to it removed.
```

File: chapters/annotations.tex (hunk 11 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2085]: There are too many open suggestions right now for also suggesting the text to go with it:
> Despite the protection level of \lstinline!P.M2!, a too must be cautious to not reveal the text with enough detail to tell that there is a reference to the class \lstinline!P.MySecret! (this is different compared to when such misuse of a protected class occurs outside of the package containing the protected class).

File: chapters/annotations.tex (hunk 7 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2082]: To me this use of origin and extent for seems like a weird style, but as stated before the specification shouldn't try to enforce a style-guide.

Comment (HansOlsson) [reply chapters/annotations.tex]: > For what it's worth, I think System Modeler does it this way as a way of giving the user control of the center of rotation used in GUI rotation operations. 

I believe Dymola introduces the origin (originally in the center) when you start rotating.

Comment (henrikt-ma) [reply chapters/annotations.tex]: Right, that should work too.

**Comment by HansOlsson** [comment]
Language group:
System modeler (Henrik): should be possible to refuse to load if not consistently used in visible part of library (negotiate with library developer). Should be seen as lookup error.

Tools should support developing such libraries similarly as other libraries without any restrictions, restrictions only apply when it is used.

**Comment by henrikt-ma** [comment]
I'd be interested in going forward with the things agreed upon at the last [phone meeting](https://github.com/modelica/ModelicaSpecification/pull/3232#issuecomment-1274794049), but it would be good if currently open conversations could be resolved first to avoid GitHub suggestion chaos.

File: chapters/annotations.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2066]: The original text used should and must according to https://www.rfc-editor.org/rfc/rfc2119 - "may not" is not included in that list. (Well, "should" was possibly borderline.) Thus, I prefer to not change this. The other words could be changed.

File: chapters/annotations.tex (hunk 5 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2069]: Similarly.

File: chapters/annotations.tex (hunk 11 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2090]: I understand the concern, but one could also just have a text saying: "There is a hidden class MySecret" in the documentation, or having `//MySecret mySecret;` in the code.

(Technically using `MySecret` in an exposed class inside the protected package doesn't necessarily imply that there's a class called `MySecret`, as the package might have `import MySecret=Modelica.Blocks.Math.Gain;`)

File: chapters/annotations.tex (hunk 4 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2066]: It is this very meaning from RFC2119 that I find inappropriate in these statements:
> **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.

I don't think that the specification should open up for using the classes in ways that contradict the protection.  This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

File: chapters/annotations.tex (hunk 5 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:2069]: Right, let's keep the discussion in the thread above.

**Comment by HansOlsson** [comment]
Phone meeting:
Tools should ideally have some way of checking this before encrypting.

The discussion focuses on two issues:
* How illegal is it to violate the rules? (Should vs must)
* Accessing text that uses protected classes.

File: chapters/annotations.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2066]: > It is this very meaning from RFC2119 that I find inappropriate in these statements:
> 
> > **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.
> 
> I don't think that the specification should open up for using the classes in ways that contradict the protection. This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

If it is an **absolute requirement**, I don't see that it makes sense to specify the behavior.
We could clarify that "The only acceptable use of class contradicting the protection is for testing the protection."

Comment (henrikt-ma) [reply chapters/annotations.tex:2066]: I agree that it would be best to not specify the behavior.  I would prefer to not clarify anything at all:
- The specification shouldn't say anything at all regarding what kind of testing tool vendors should perform in order to check their compliance with the specification.  On GitHub, we _could_ provide a test collection that can help tool vendors check their compliance, but it is not the specification's job to point out such resources.
- Instead of hinting that library authors perhaps should worry about how tools behave if the library violates the specification, we should explain the importance of library authors verifying that a tool meets their expectations when it comes to supporting correct `protection`-annotation behavior (for valid cases), before they allow their library to be decrypted by the tool.

Comment (HansOlsson) [reply chapters/annotations.tex:2066]: > I agree that it would be best to not specify the behavior.

I don't. The reason is that such issues occur does occur in practice.
However, if we don't want to specify the behavior it seems odd to make it an absolute requirement.

**Comment by HansOlsson** [comment]
I'm not seeing any sensible solution for the issue in practice, except the currently proposed one.

The issue is:
* A library has been encrypted - and for some reason some model has a diagram that uses a model that shouldn't be exposed.
* _Without requiring_ that tools instantiate _all_ models in the library to ensure that they satisfy the requirements before showing anything.

**Review by eshmoylova — CHANGES_REQUESTED** [review]

**Review by eshmoylova — APPROVED** [review]

File: chapters/annotations.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2066]: One point is at _what level_ it shall be rejected.

If you encrypt library of 100 models, does a tool have to do some detailed(*) check that _all of them_ satisfy this before doing anything?

I would say not - and thus I see two realistic alternatives:
- Restrict what is shown for that model, but show other models
- Don't show the contents of that model, but show other models

Both are ok to me; but neither is consistent with viewing it as an absolute requirement.
Note that this may get complicated if the problematic model is used internally in the library.

*: There are also some non-detailed checks. E.g., if a tool cannot parse the library it doesn't make sense to present anything. However, as far as I understand SEMLA is designed so that even parsing of sub-parts can be delayed (similarly as for normal Modelica).

**Comment by HansOlsson** [comment]
Language group: 
Favor 'should': Hans, Stephan, Elena
Against ('may not'):
Abstain: Dag, Quentin, Gerd, Markus

File: chapters/annotations.tex (hunk 4 above)
Comment (HansOlsson) [reply chapters/annotations.tex:2066]: This was resolved at the latest phone meeting.

**Comment by HansOlsson** [comment]
> I'd be interested in going forward with the things agreed upon at the last [phone meeting](https://github.com/modelica/ModelicaSpecification/pull/3232#issuecomment-1274794049), but it would be good if currently open conversations could be resolved first to avoid GitHub suggestion chaos.

Everything should now be resolved.

**Comment by henrikt-ma** [comment]
> Language group: Favor 'should': Hans, Stephan, Elena Against ('may not'): Abstain: Dag, Quentin, Gerd, Markus

As this doesn't make any sense to us at Wolfram (Quentin may not have taken part of the discussions), I don't foresee System Modeler ever complying with the current formulation.
